newsroom/users/views.py

```diff
@@ -56,196 +75,241 @@ def get_settings_data():
-    return jsonify(users), 200
+    return Response(users, 200, ())
 
 
-@blueprint.route("/users/new", methods=["POST"])
+@users_endpoints.endpoint("/users/new", methods=["POST"])
 @account_manager_or_company_admin_only
-async def create():
+async def create(request: Request):
     form = await UserForm.create_form()
+
     if await form.validate():
         if not _is_email_address_valid(form.email.data):
-            return jsonify({"email": [gettext("Email address is already in use")]}), 400
+            return Response({"email": [gettext("Email address is already in use")]}, 400, ())
 
-        new_user = get_updates_from_form(form, on_create=True)
-        user_is_company_admin = is_current_user_company_admin()
-        if user_is_company_admin:
-            company = get_company()
-            if company is None:
-                abort(401)
+        creation_data = get_updates_from_form(form, on_create=True)
+        new_user = UserResourceModel(**creation_data, id=ObjectId())
+
+        if is_current_user_company_admin():
+            company_from_admin = await get_company_from_user_or_session()
+            if not company_from_admin:
+                return await request.abort(401)
 
             # Make sure this new user is associated with ``company`` and as a ``PUBLIC`` user
-            new_user["company"] = company["_id"]
-            new_user["user_type"] = UserRole.PUBLIC.value
+            new_user.company = company_from_admin.id
+            new_user.user_type = UserRole.PUBLIC
         elif form.company.data:
-            new_user["company"] = ObjectId(form.company.data)
-        elif new_user["user_type"] != "administrator":
-            return (
-                jsonify({"company": [gettext("Company is required for non administrators")]}),
-                400,
-            )
+            new_user.company = ObjectId(form.company.data)
+        elif new_user.user_type != "administrator":
+            return Response({"company": [gettext("Company is required for non administrators")]}, 400, ())
 
-        # Flask form won't accept default value if any form data was passed in the request.
-        # So, we need to set this explicitly here.
-        new_user["receive_email"] = True
-        new_user["receive_app_notifications"] = True
+        new_user.receive_email = True
+        new_user.receive_app_notifications = True
 
-        company = get_company(new_user)
-        auth_provider = get_company_auth_provider(company)
+        company = await get_company_from_user_or_session(new_user)
+        if company:
+            auth_provider = get_company_auth_provider(company.model_dump(by_alias=True))
 
-        if auth_provider.features["verify_email"]:
-            add_token_data(new_user)
+            if auth_provider.features["verify_email"]:
+                add_token_data(new_user)
 
-        ids = get_resource_service("users").post([new_user])
+        ids = await UsersService().create([new_user])
```

Comment:
Sorry, I missed this one. Would need to wrap `service.create` and `service.update` in a try...except to catch and return validation errors to the browser.

See an example [here](https://github.com/superdesk/newsroom-core/blob/async/newsroom/companies/views.py#L84)

Reply:
done.